### ISSUE TITLE
delta.checkpointInterval breaks writes on 0.22.0

### ISSUE THREAD

**Issue by echai58** [delta-io/delta-rs#3030]
# Environment

**Delta-rs version**: 0.22.0

**Binding**: python

***
# Bug

**What happened**: I am trying to upgrade to the latest release of delta-rs and it seems to introduce a breaking bug when using `delta.checkpointInterval`.

**What you expected to happen**: checkpointInterval continues to work as before, and I can write safely.

**How to reproduce it**: This introduces a breaking bug in both pyarrow and rust writer engines. In pyarrow, it does not overwrite successfully (two rows in output), and in rust, it panics.

```
from deltalake import DeltaTable, write_deltalake
import pandas as pd 

write_deltalake(
    "test",
    pd.DataFrame(
        {
            "a": ["a"],
            "b": [1],
        }
    ),
    configuration={
        "delta.checkpointInterval": "5"
    }
)

for i in range(4):
    write_deltalake(
        "test",
            pd.DataFrame(
                {
                    "a": ["a"],
                    "b": [i],
                }
            ),
        mode="overwrite",
    )
```

At this point, the delta table looks correct, e.g.:
![image](https://github.com/user-attachments/assets/46e0b9fb-d55e-474f-82dd-9b0731f0a163)

But, on the next write, when using `rust` writer engine, we get the following exception:
```
PanicException: called `Result::unwrap()` on an `Err` value: DeletionVector("Unknown storage format: ''.")
```

and on pyarrow, it manifests itself with an incorrect overwrite:
![image](https://github.com/user-attachments/assets/575bd3fd-9415-487d-8238-a663fd76ef2e)
with two rows showing up.

**Comment by ion-elgreco** [comment]
This is delta-kernel-rs code where it fails

**Comment by TinoSM** [comment]
It also happens when doing overwrites from polars (simple table overwrites)

To me it happens in every overwrite, was trying to see if its something for polars to adapt or delta-rs

**Comment by ion-elgreco** [comment]
Very unclear why this now fails

Trace:
```
thread '<unnamed>' panicked at crates/core/src/kernel/snapshot/log_data.rs:98:55:
called `Result::unwrap()` on an `Err` value: DeletionVector("Unknown storage format: ''.")
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by rtyler** [comment]
@echai58 thanks for the bug report and the reproduction case! I'll spin up a branch with this test case

**Comment by rtyler** [comment]
The checkpoint is being re-read as having an empty DeletionVector

**Comment by echai58** [comment]
@rtyler thanks for investigating - does it help narrow it down that the error is only raised when overwriting with `engine=rust`, vs `engine=pyarrow` just seemingly ignoring the existing data, and appending the new data even when `mode=overwrite`?

**Comment by scovich** [comment]
Looking at this output:
```
Some(DeletionVector { storage_type: StringArray
[
  "",
], path_or_inline_dv: StringArray
```
The empty string is not a valid `storage_type` -- according to the [Delta spec](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#deletion-vector-descriptor-schema):
> Legal options are: ['u', 'i', 'p'].

It looks to me like some code somewhere is translating "missing" to "default value" instead of NULL: `0`, `""`, etc. Interestingly, the one nullable DV descriptor field _does_ come through as NULL:
```
], offset: Some(PrimitiveArray<Int32>
[
  null,
]) })
```
... which makes me suspect some code somewhere is not gracefully handling the case of a non-nullable struct field inside a nullable struct.

Do we know what is physically written in the parquet and/or json files? That would tell us whether it's a read path or write path issue.

**Comment by scovich** [comment]
Or, even more likely, these were arrow reads that ignored a null mask and picked up whatever (zero-initialized bytes) happened to be in the corresponding data column.

**Comment by ldacey** [comment]
> To me this error also happens when doing overwrites from polars (simple table overwrites, appends works fine), without checkpoint involved, couldn't isolate the issue, but just saying it could not be 100% related with checkpointing

Yeah - I upgraded last night and today I found a lot of my tasks failed with this error but I have a separate checkpoint task which hasn't run yet. Also, it has not been consistently failing. Each of the red tasks have the same DeletionVector error but then it succeeds later.

<img width="317" alt="image" src="https://github.com/user-attachments/assets/a25f6540-7e4f-4cc4-aed1-a70fbe02a687">

**Comment by echai58** [comment]
@rtyler i upgraded to 0.22.2 and am still experiencing incorrect behavior. it no longer panics, but in both the `rust` and `pyarrow` engine, overwrite is not overwriting correctly:

```
from deltalake import DeltaTable, write_deltalake
import pandas as pd 

write_deltalake(
    "test",
    pd.DataFrame(
        {
            "a": ["a"],
            "b": [100],
        }
    ),
)
DeltaTable("test").create_checkpoint()
write_deltalake(
    "test",
    pd.DataFrame(
        {
            "a": ["a"],
            "b": [50],
        }
    ),
    # engine="pyarrow",
    mode="overwrite"
)
```

the table looks like:
![image](https://github.com/user-attachments/assets/0e3fd689-1f76-4a4f-b7a3-e00bf9591465)

**Comment by rtyler** [comment]
I'm investigating this further, quite a confusing issue since it seems to be directly related to the checkpoint :confused:

**Comment by ldacey** [comment]
> @rtyler i upgraded to 0.22.2 and am still experiencing incorrect behavior. it no longer panics, but in both the `rust` and `pyarrow` engine, overwrite is not overwriting correctly:
> 
> ```
> from deltalake import DeltaTable, write_deltalake
> import pandas as pd 
> 
> write_deltalake(
>     "test",
>     pd.DataFrame(
>         {
>             "a": ["a"],
>             "b": [100],
>         }
>     ),
> )
> DeltaTable("test").create_checkpoint()
> write_deltalake(
>     "test",
>     pd.DataFrame(
>         {
>             "a": ["a"],
>             "b": [50],
>         }
>     ),
>     # engine="pyarrow",
>     mode="overwrite"
> )
> ```
> 
> the table looks like: ![image](https://private-user-images.githubusercontent.com/56415623/391587568-0e3fd689-1f76-4a4f-b7a3-e00bf9591465.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMyNDc4MjUsIm5iZiI6MTczMzI0NzUyNSwicGF0aCI6Ii81NjQxNTYyMy8zOTE1ODc1NjgtMGUzZmQ2ODktMWY3Ni00YTRmLWI3YTMtZTAwYmY5NTkxNDY1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjAzVDE3Mzg0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU3ODA0ZTljNTJiYTJmYWExZTRhN2IwNjM0NDAyYjA1NmM0M2VhZjc4NDU4ZmQ0NzBlNzg0MDQ3Nzk2MzEwYmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.bKwOCKGlCQUWG4vSHSQcwd-OcLu_hU_yAQsOEoL_Buc)

Thanks for this comment. This explains an issue I was facing with duplicates in my gold layer which I could not reproduce locally (because I was not running the create_checkpoint())

**Comment by ion-elgreco** [comment]
Fyi, I've yanked all the 0.22.X releases
